File: mac/Sources/VibeBoardKit/VKA1/VKA1.swift
import CryptoKit
import Foundation

public enum VKA1Error: Error, Equatable, Sendable {
    case invalidHeader
    case invalidKind
    case invalidPixelFormat
    case invalidEncoding
    case invalidDimensions
    case invalidFrameCount
    case invalidDuration
    case invalidRange
    case invalidHash
    case invalidRLE
    case limitExceeded
}

public enum VKA1Kind: UInt8, Sendable {
    case image = 1
    case animation = 2
    case glyphBitmap = 3
}

public enum VKA1FrameEncoding: UInt8, Sendable {
    case raw = 0
    case rowRLE = 1
}

public struct VKA1Limits: Equatable, Sendable {
    public let maxFrames: UInt16
    public let minFrameDurationMS: UInt16
    public let maxFrameDurationMS: UInt16
    public let maxContainerBytes: UInt32
    public let maxDecodedBytes: UInt32

    public init(maxFrames: UInt16, minFrameDurationMS: UInt16, maxFrameDurationMS: UInt16, maxContainerBytes: UInt32, maxDecodedBytes: UInt32) {
        self.maxFrames = maxFrames
        self.minFrameDurationMS = minFrameDurationMS
        self.maxFrameDurationMS = maxFrameDurationMS
        self.maxContainerBytes = maxContainerBytes
        self.maxDecodedBytes = maxDecodedBytes
    }
}

public struct VKA1SourceFrame: Equatable, Sendable {
    public let pixels: [UInt16]
    public let durationMS: UInt16

    public init(pixels: [UInt16], durationMS: UInt16) {
        self.pixels = pixels
        self.durationMS = durationMS
    }
}

public struct VKA1Frame: Equatable, Sendable {
    public let encoding: VKA1FrameEncoding
    public let durationMS: UInt16
    public let pixels: [UInt16]
}

public struct VKA1Container: Equatable, Sendable {
    public let kind: VKA1Kind
    public let width: UInt16
    public let height: UInt16
    public let frames: [VKA1Frame]
    public let sha256: String
}

public enum VKA1Codec {
    public static func encode(kind: VKA1Kind, width: UInt16, height: UInt16, frames: [VKA1SourceFrame], limits: VKA1Limits) throws -> Data {
        let pixelCount = try checkedGeometry(width: width, height: height, limits: limits)
        guard !frames.isEmpty, frames.count <= Int(limits.maxFrames), frames.count <= Int(UInt16.max) else { throw VKA1Error.invalidFrameCount }
        try validateKind(kind, frames: frames, limits: limits)

        var encoded: [(Data, VKA1FrameEncoding, UInt16)] = []
        encoded.reserveCapacity(frames.count)
        var encodingBits: UInt8 = 0
        for frame in frames {
            guard frame.pixels.count == pixelCount else { throw VKA1Error.invalidDimensions }
            let raw = rawData(frame.pixels)
            let rle = try rowRLE(frame.pixels, width: Int(width), height: Int(height))
            let choice: (Data, VKA1FrameEncoding) = rle.count < raw.count ? (rle, .rowRLE) : (raw, .raw)
            encodingBits |= choice.1 == .raw ? 1 : 2
            encoded.append((choice.0, choice.1, frame.durationMS))
        }

        let (tableBytes, tableOverflow) = frames.count.multipliedReportingOverflow(by: 12)
        let (headerBytes, headerOverflow) = 56.addingReportingOverflow(tableBytes)
        guard !tableOverflow, !headerOverflow, headerBytes <= Int(UInt16.max) else { throw VKA1Error.limitExceeded }
        var total = headerBytes
        for frame in encoded {
            let (next, overflow) = total.addingReportingOverflow(frame.0.count)
            guard !overflow, next <= Int(limits.maxContainerBytes), next <= Int(UInt32.max) else { throw VKA1Error.limitExceeded }
            total = next
        }

        var data = Data()
        data.reserveCapacity(total)
        data.append(contentsOf: [0x56, 0x4b, 0x41, 0x31, kind.rawValue, 1, encodingBits, 0])
        appendLE(width, to: &data)
        appendLE(height, to: &data)
        appendLE(UInt16(frames.count), to: &data)
        appendLE(UInt16(headerBytes), to: &data)
        appendLE(UInt32(pixelCount * 2), to: &data)
        appendLE(UInt32(total), to: &data)
        data.append(Data(repeating: 0, count: 32))

        var offset = headerBytes
        for frame in encoded {
            appendLE(UInt32(offset), to: &data)
            appendLE(UInt32(frame.0.count), to: &data)
            appendLE(frame.2, to: &data)
            data.append(frame.1.rawValue)
            data.append(0)
            offset += frame.0.count
        }
        for frame in encoded { data.append(frame.0) }
        let digest = SHA256.hash(data: data)
        data.replaceSubrange(24..<56, with: Data(digest))
        _ = try decode(data, limits: limits)
        return data
    }

    public static func decode(_ data: Data, limits: VKA1Limits) throws -> VKA1Container {
        guard data.count >= 68, data.count <= Int(limits.maxContainerBytes), data.prefix(4) == Data([0x56, 0x4b, 0x41, 0x31]) else { throw VKA1Error.invalidHeader }
        guard let kind = VKA1Kind(rawValue: data[4]) else { throw VKA1Error.invalidKind }
        guard data[5] == 1, data[7] == 0, data[6] & ~UInt8(3) == 0 else { throw VKA1Error.invalidEncoding }
        let width = read16(data, 8), height = read16(data, 10), frameCount = read16(data, 12), headerBytes = read16(data, 14)
        let decodedBytes = read32(data, 16), totalBytes = read32(data, 20)
        let pixelCount = try checkedGeometry(width: width, height: height, limits: limits)
        guard frameCount > 0, frameCount <= limits.maxFrames else { throw VKA1Error.invalidFrameCount }
        let (aggregateDecodedBytes, aggregateOverflow) = Int(decodedBytes).multipliedReportingOverflow(by: Int(frameCount))
        guard !aggregateOverflow, aggregateDecodedBytes <= Int(limits.maxDecodedBytes) else {
            throw VKA1Error.limitExceeded
        }
        let expectedHeader = 56 + Int(frameCount) * 12
        guard Int(headerBytes) == expectedHeader, totalBytes == UInt32(data.count), decodedBytes == UInt32(pixelCount * 2) else { throw VKA1Error.invalidHeader }
        var zeroed = data
        let expectedHash = data[24..<56]
        zeroed.replaceSubrange(24..<56, with: repeatElement(UInt8(0), count: 32))
        guard Data(SHA256.hash(data: zeroed)) == expectedHash else { throw VKA1Error.invalidHash }

        var entries: [(Int, Int, UInt16, VKA1FrameEncoding)] = []
        var expectedOffset = expectedHeader
        var union: UInt8 = 0
        for index in 0..<Int(frameCount) {
            let base = 56 + index * 12
            let offset = Int(read32(data, base)), length = Int(read32(data, base + 4)), duration = read16(data, base + 8)
            guard let encoding = VKA1FrameEncoding(rawValue: data[base + 10]), data[base + 11] == 0 else { throw VKA1Error.invalidEncoding }
            guard offset == expectedOffset, length > 0 else { throw VKA1Error.invalidRange }
            let (end, overflow) = offset.addingReportingOverflow(length)
            guard !overflow, end <= data.count else { throw VKA1Error.invalidRange }
            expectedOffset = end
            union |= encoding == .raw ? 1 : 2
            entries.append((offset, length, duration, encoding))
        }
        guard expectedOffset == data.count, union == data[6] else { throw VKA1Error.invalidRange }
        try validateDurations(kind, entries.map(\.2), limits: limits)

        var frames: [VKA1Frame] = []
        frames.reserveCapacity(entries.count)
        for entry in entries {
            let payload = data[entry.0..<(entry.0 + entry.1)]
            let pixels: [UInt16]
            switch entry.3 {
            case .raw:
                guard payload.count == pixelCount * 2 else { throw VKA1Error.invalidRange }
                pixels = stride(from: payload.startIndex, to: payload.endIndex, by: 2).map { UInt16(payload[$0]) | UInt16(payload[$0 + 1]) << 8 }
                let canonicalRLEBytes = try rowRLE(pixels, width: Int(width), height: Int(height)).count
                guard canonicalRLEBytes >= payload.count else { throw VKA1Error.invalidEncoding }
            case .rowRLE:
                pixels = try decodeRLE(payload, width: Int(width), height: Int(height))
                guard payload.count < pixelCount * 2 else { throw VKA1Error.invalidEncoding }
            }
            frames.append(VKA1Frame(encoding: entry.3, durationMS: entry.2, pixels: pixels))
        }
        return VKA1Container(kind: kind, width: width, height: height, frames: frames, sha256: expectedHash.map { String(format: "%02x", $0) }.joined())
    }

    private static func checkedGeometry(width: UInt16, height: UInt16, limits: VKA1Limits) throws -> Int {
        guard (1...428).contains(width), (1...142).contains(height) else { throw VKA1Error.invalidDimensions }
        let pixels = Int(width) * Int(height)
        guard pixels <= Int(UInt32.max / 2), pixels * 2 <= Int(limits.maxDecodedBytes) else { throw VKA1Error.limitExceeded }
        return pixels
    }

    private static func validateKind(_ kind: VKA1Kind, frames: [VKA1SourceFrame], limits: VKA1Limits) throws {
        try validateDurations(kind, frames.map(\.durationMS), limits: limits)
    }

    private static func validateDurations(_ kind: VKA1Kind, _ durations: [UInt16], limits: VKA1Limits) throws {
        switch kind {
        case .image, .glyphBitmap:
            guard durations.count == 1, durations[0] == 0 else { throw VKA1Error.invalidDuration }
        case .animation:
            guard durations.allSatisfy({ $0 > 0 && $0 >= limits.minFrameDurationMS && $0 <= limits.maxFrameDurationMS }) else { throw VKA1Error.invalidDuration }
        }
    }

    private static func rawData(_ pixels: [UInt16]) -> Data {
        var data = Data(); data.reserveCapacity(pixels.count * 2)
        for pixel in pixels { appendLE(pixel, to: &data) }
        return data
    }

    private static func rowRLE(_ pixels: [UInt16], width: Int, height: Int) throws -> Data {
        guard width > 0, height > 0, pixels.count == width * height, width <= Int(UInt16.max) else { throw VKA1Error.invalidDimensions }
        var data = Data()
        for row in 0..<height {
            var column = 0
            while column < width {
                let pixel = pixels[row * width + column]
                var end = column + 1
                while end < width, pixels[row * width + end] == pixel { end += 1 }
                guard let count = UInt16(exactly: end - column) else { throw VKA1Error.limitExceeded }
                appendLE(count, to: &data); appendLE(pixel, to: &data)
                column = end
            }
        }
        return data
    }

    private static func decodeRLE(_ payload: Data.SubSequence, width: Int, height: Int) throws -> [UInt16] {
        var pixels: [UInt16] = []; pixels.reserveCapacity(width * height)
        var index = payload.startIndex
        for _ in 0..<height {
            var columns = 0
            var prior: UInt16?
            while columns < width {
                guard index + 4 <= payload.endIndex else { throw VKA1Error.invalidRLE }
                let count = Int(UInt16(payload[index]) | UInt16(payload[index + 1]) << 8)
                let pixel = UInt16(payload[index + 2]) | UInt16(payload[index + 3]) << 8
                guard count > 0, count <= width - columns, prior != pixel else { throw VKA1Error.invalidRLE }
                pixels.append(contentsOf: repeatElement(pixel, count: count))
                columns += count; prior = pixel; index += 4
            }
        }
        guard index == payload.endIndex else { throw VKA1Error.invalidRLE }
        return pixels
    }

    private static func appendLE(_ value: UInt16, to data: inout Data) { data.append(UInt8(truncatingIfNeeded: value)); data.append(UInt8(truncatingIfNeeded: value >> 8)) }
    private static func appendLE(_ value: UInt32, to data: inout Data) { for shift in stride(from: 0, through: 24, by: 8) { data.append(UInt8(truncatingIfNeeded: value >> UInt32(shift))) } }
    private static func read16(_ data: Data, _ offset: Int) -> UInt16 { UInt16(data[offset]) | UInt16(data[offset + 1]) << 8 }
    private static func read32(_ data: Data, _ offset: Int) -> UInt32 { UInt32(data[offset]) | UInt32(data[offset + 1]) << 8 | UInt32(data[offset + 2]) << 16 | UInt32(data[offset + 3]) << 24 }
}

File: win/src/VibeBoardKit/VKA1/VKA1.cs
using System;
using System.Collections.Generic;
using System.Security.Cryptography;
using System.Text;

namespace VibeBoardKit.VKA1;

public class VKA1Exception : Exception
{
    public VKA1Exception(string message) : base(message) { }
}

public enum VKA1Kind : byte
{
    Image = 1,
    Animation = 2,
    GlyphBitmap = 3
}

public enum VKA1FrameEncoding : byte
{
    Raw = 0,
    RowRLE = 1
}

public sealed class VKA1Limits
{
    public ushort MaxFrames { get; }
    public ushort MinFrameDurationMS { get; }
    public ushort MaxFrameDurationMS { get; }
    public uint MaxContainerBytes { get; }
    public uint MaxDecodedBytes { get; }

    public VKA1Limits(ushort maxFrames, ushort minFrameDurationMS, ushort maxFrameDurationMS,
        uint maxContainerBytes, uint maxDecodedBytes)
    {
        MaxFrames = maxFrames;
        MinFrameDurationMS = minFrameDurationMS;
        MaxFrameDurationMS = maxFrameDurationMS;
        MaxContainerBytes = maxContainerBytes;
        MaxDecodedBytes = maxDecodedBytes;
    }
}

public sealed class VKA1SourceFrame
{
    public ushort[] Pixels { get; }
    public ushort DurationMS { get; }
    public VKA1SourceFrame(ushort[] pixels, ushort durationMS)
    { Pixels = pixels; DurationMS = durationMS; }
}

public sealed class VKA1Frame
{
    public VKA1FrameEncoding Encoding { get; }
    public ushort DurationMS { get; }
    public ushort[] Pixels { get; }
    public VKA1Frame(VKA1FrameEncoding encoding, ushort durationMS, ushort[] pixels)
    { Encoding = encoding; DurationMS = durationMS; Pixels = pixels; }
}

public sealed class VKA1Container
{
    public VKA1Kind Kind { get; }
    public ushort Width { get; }
    public ushort Height { get; }
    public List<VKA1Frame> Frames { get; }
    public string SHA256 { get; }

    public VKA1Container(VKA1Kind kind, ushort width, ushort height, List<VKA1Frame> frames, string sha256)
    { Kind = kind; Width = width; Height = height; Frames = frames; SHA256 = sha256; }
}

/// <summary>
/// Codec for the VKA1 binary asset container format.
/// This is the device-side asset format for images and animations.
/// </summary>
public static class VKA1Codec
{
    public static byte[] Encode(VKA1Kind kind, ushort width, ushort height,
        List<VKA1SourceFrame> frames, VKA1Limits limits)
    {
        int pixelCount = CheckedGeometry(width, height, limits);
        if (frames.Count == 0 || frames.Count > limits.MaxFrames || frames.Count > ushort.MaxValue)
            throw new VKA1Exception("Invalid frame count");
        ValidateKind(kind, frames, limits);

        var encoded = new List<(byte[] data, VKA1FrameEncoding encoding, ushort durationMS)>(frames.Count);
        byte encodingBits = 0;

        foreach (var frame in frames)
        {
            if (frame.Pixels.Length != pixelCount)
                throw new VKA1Exception("Invalid dimensions");
            byte[] raw = RawData(frame.Pixels);
            byte[] rle = RowRLE(frame.Pixels, width, height);
            var choice = rle.Length < raw.Length ? (rle, VKA1FrameEncoding.RowRLE) : (raw, VKA1FrameEncoding.Raw);
            encodingBits |= (byte)(choice.Item2 == VKA1FrameEncoding.Raw ? 1 : 2);
            encoded.Add((choice.Item1, choice.Item2, frame.DurationMS));
        }

        int headerBytes = 56 + frames.Count * 12;
        if (headerBytes > ushort.MaxValue)
            throw new VKA1Exception("Limit exceeded: header");

        long total = headerBytes;
        foreach (var frame in encoded)
        {
            total += frame.data.Length;
            if (total > limits.MaxContainerBytes || total > uint.MaxValue)
                throw new VKA1Exception("Limit exceeded: container");
        }

        var data = new List<byte>((int)total);
        // Magic
        data.AddRange(new byte[] { 0x56, 0x4b, 0x41, 0x31, (byte)kind, 1, encodingBits, 0 });
        AppendLE(data, width);
        AppendLE(data, height);
        AppendLE(data, (ushort)frames.Count);
        AppendLE(data, (ushort)headerBytes);
        AppendLE(data, (uint)(pixelCount * 2));
        AppendLE(data, (uint)total);
        // SHA256 placeholder (32 zeros)
        data.AddRange(new byte[32]);

        // Frame table
        int offset = headerBytes;
        foreach (var frame in encoded)
        {
            AppendLE(data, (uint)offset);
            AppendLE(data, (uint)frame.data.Length);
            AppendLE(data, frame.durationMS);
            data.Add((byte)frame.encoding);
            data.Add(0);
            offset += frame.data.Length;
        }

        // Frame data
        foreach (var frame in encoded)
            data.AddRange(frame.data);

        // Compute SHA256 with hash field zeroed
        var hashData = data.ToArray();
        var sha256 = SHA256.HashData(hashData);
        // Write hash into bytes 24..56
        Array.Copy(sha256, 0, hashData, 24, 32);

        // Self-verify
        Decode(hashData, limits);
        return hashData;
    }

    public static VKA1Container Decode(byte[] data, VKA1Limits limits)
    {
        if (data.Length < 68 || data.Length > limits.MaxContainerBytes)
            throw new VKA1Exception("Invalid header");
        if (data[0] != 0x56 || data[1] != 0x4b || data[2] != 0x41 || data[3] != 0x31)
            throw new VKA1Exception("Invalid header: magic");

        var kind = (VKA1Kind)data[4];
        if (!Enum.IsDefined(typeof(VKA1Kind), kind))
            throw new VKA1Exception("Invalid kind");
        if (data[5] != 1 || data[7] != 0 || (data[6] & ~3) != 0)
            throw new VKA1Exception("Invalid encoding");

        ushort width = Read16(data, 8);
        ushort height = Read16(data, 10);
        ushort frameCount = Read16(data, 12);
        ushort headerBytes = Read16(data, 14);
        uint decodedBytes = Read32(data, 16);
        uint totalBytes = Read32(data, 20);

        int pixelCount = CheckedGeometry(width, height, limits);
        if (frameCount == 0 || frameCount > limits.MaxFrames)
            throw new VKA1Exception("Invalid frame count");

        long aggregateDecoded = (long)decodedBytes * frameCount;
        if (aggregateDecoded > limits.MaxDecodedBytes)
            throw new VKA1Exception("Limit exceeded: decoded");

        int expectedHeader = 56 + frameCount * 12;
        if (headerBytes != expectedHeader || totalBytes != (uint)data.Length || decodedBytes != (uint)(pixelCount * 2))
            throw new VKA1Exception("Invalid header: size mismatch");

        // Verify SHA256
        byte[] expectedHash = new byte[32];
        Array.Copy(data, 24, expectedHash, 0, 32);
        var zeroed = (byte[])data.Clone();
        Array.Clear(zeroed, 24, 32);
        var actualHash = SHA256.HashData(zeroed);
        if (!actualHash.AsSpan().SequenceEqual(expectedHash))
            throw new VKA1Exception("Invalid hash");

        // Parse frame table
        var entries = new List<(int offset, int length, ushort duration, VKA1FrameEncoding encoding)>();
        int expectedOffset = expectedHeader;
        byte union = 0;

        for (int i = 0; i < frameCount; i++)
        {
            int base_ = 56 + i * 12;
            int offset = (int)Read32(data, base_);
            int length = (int)Read32(data, base_ + 4);
            ushort duration = Read16(data, base_ + 8);
            var encoding = (VKA1FrameEncoding)data[base_ + 10];
            if (!Enum.IsDefined(typeof(VKA1FrameEncoding), encoding))
                throw new VKA1Exception("Invalid encoding");
            if (data[base_ + 11] != 0)
                throw new VKA1Exception("Invalid encoding: reserved");

            if (offset != expectedOffset || length <= 0)
                throw new VKA1Exception("Invalid range");
            int end = offset + length;
            if (end > data.Length)
                throw new VKA1Exception("Invalid range: end");
            expectedOffset = end;
            union |= (byte)(encoding == VKA1FrameEncoding.Raw ? 1 : 2);
            entries.Add((offset, length, duration, encoding));
        }

        if (expectedOffset != data.Length || union != data[6])
            throw new VKA1Exception("Invalid range: union");

        var durations = new ushort[entries.Count];
        for (int i = 0; i < entries.Count; i++)
            durations[i] = entries[i].duration;
        ValidateDurations(kind, durations, limits);

        var frames = new List<VKA1Frame>(entries.Count);
        foreach (var entry in entries)
        {
            var payload = new byte[entry.length];
            Array.Copy(data, entry.offset, payload, 0, entry.length);
            ushort[] pixels;

            if (entry.encoding == VKA1FrameEncoding.Raw)
            {
                if (payload.Length != pixelCount * 2)
                    throw new VKA1Exception("Invalid range: raw size");
                pixels = new ushort[pixelCount];
                for (int i = 0; i < pixelCount; i++)
                    pixels[i] = (ushort)(payload[i * 2] | (payload[i * 2 + 1] << 8));

                // Canonical RLE must be >= raw (device chose raw, so RLE must not be smaller)
                int canonicalRLELength = RowRLE(pixels, width, height).Length;
                if (canonicalRLELength < payload.Length)
                    throw new VKA1Exception("Invalid encoding: canonical RLE smaller than raw");
            }
            else
            {
                pixels = DecodeRLE(payload, width, height);
                if (payload.Length >= pixelCount * 2)
                    throw new VKA1Exception("Invalid encoding: RLE not smaller");
            }

            frames.Add(new VKA1Frame(entry.encoding, entry.duration, pixels));
        }

        var hashStr = new StringBuilder(64);
        foreach (byte b in expectedHash)
            hashStr.Append(b.ToString("x2"));

        return new VKA1Container(kind, width, height, frames, hashStr.ToString());
    }

    private static int CheckedGeometry(ushort width, ushort height, VKA1Limits limits)
    {
        if (width < 1 || width > 428 || height < 1 || height > 142)
            throw new VKA1Exception("Invalid dimensions");
        int pixels = width * height;
        if (pixels > uint.MaxValue / 2 || (long)pixels * 2 > limits.MaxDecodedBytes)
            throw new VKA1Exception("Limit exceeded: decoded");
        return pixels;
    }

    private static void ValidateKind(VKA1Kind kind, List<VKA1SourceFrame> frames, VKA1Limits limits)
    {
        var durations = new ushort[frames.Count];
        for (int i = 0; i < frames.Count; i++)
            durations[i] = frames[i].DurationMS;
        ValidateDurations(kind, durations, limits);
    }

    private static void ValidateDurations(VKA1Kind kind, ushort[] durations, VKA1Limits limits)
    {
        switch (kind)
        {
            case VKA1Kind.Image:
            case VKA1Kind.GlyphBitmap:
                if (durations.Length != 1 || durations[0] != 0)
                    throw new VKA1Exception("Invalid duration");
                break;
            case VKA1Kind.Animation:
                foreach (var d in durations)
                {
                    if (d == 0 || d < limits.MinFrameDurationMS || d > limits.MaxFrameDurationMS)
                        throw new VKA1Exception("Invalid duration");
                }
                break;
        }
    }

    private static byte[] RawData(ushort[] pixels)
    {
        var data = new byte[pixels.Length * 2];
        for (int i = 0; i < pixels.Length; i++)
        {
            data[i * 2] = (byte)(pixels[i] & 0xFF);
            data[i * 2 + 1] = (byte)((pixels[i] >> 8) & 0xFF);
        }
        return data;
    }

    private static byte[] RowRLE(ushort[] pixels, ushort width, ushort height)
    {
        var data = new List<byte>();
        for (int row = 0; row < height; row++)
        {
            int column = 0;
            while (column < width)
            {
                ushort pixel = pixels[row * width + column];
                int end = column + 1;
                while (end < width && pixels[row * width + end] == pixel)
                    end++;
                int count = end - column;
                if (count > ushort.MaxValue)
                    throw new VKA1Exception("Limit exceeded: RLE count");
                AppendLE(data, (ushort)count);
                AppendLE(data, pixel);
                column = end;
            }
        }
        return data.ToArray();
    }

    private static ushort[] DecodeRLE(byte[] payload, ushort widthU, ushort heightU)
    {
        int width = widthU;
        int height = heightU;
        var pixels = new List<ushort>(width * height);
        int index = 0;

        for (int row = 0; row < height; row++)
        {
            int columns = 0;
            ushort? prior = null;
            while (columns < width)
            {
                if (index + 4 > payload.Length)
                    throw new VKA1Exception("Invalid RLE: truncated");
                int count = payload[index] | (payload[index + 1] << 8);
                ushort pixel = (ushort)(payload[index + 2] | (payload[index + 3] << 8));

                if (count <= 0 || count > width - columns)
                    throw new VKA1Exception("Invalid RLE: count");
                if (prior.HasValue && prior.Value == pixel)
                    throw new VKA1Exception("Invalid RLE: consecutive same pixel");

                for (int i = 0; i < count; i++)
                    pixels.Add(pixel);
                columns += count;
                prior = pixel;
                index += 4;
            }
        }

        if (index != payload.Length)
            throw new VKA1Exception("Invalid RLE: trailing data");

        return pixels.ToArray();
    }

    private static void AppendLE(List<byte> data, ushort value)
    {
        data.Add((byte)(value & 0xFF));
        data.Add((byte)((value >> 8) & 0xFF));
    }

    private static void AppendLE(List<byte> data, uint value)
    {
        data.Add((byte)(value & 0xFF));
        data.Add((byte)((value >> 8) & 0xFF));
        data.Add((byte)((value >> 16) & 0xFF));
        data.Add((byte)((value >> 24) & 0xFF));
    }

    private static ushort Read16(byte[] data, int offset)
    {
        return (ushort)(data[offset] | (data[offset + 1] << 8));
    }

    private static uint Read32(byte[] data, int offset)
    {
        return (uint)(data[offset] | (data[offset + 1] << 8) |
            (data[offset + 2] << 16) | (data[offset + 3] << 24));
    }
}
